Support Generic PostgREST 9  Claims

olirice opened the pull request:
## What kind of change does this PR introduce?
- Removes hard coded use of `authenticated` role
- Replaces `realtime.subscription` email, and role fields with generic jwt claims jsonb field

at minimum, that the claims must contain a top level key named role e.g. `{"role": "authenticated"}`, which is verified by a trigger to be a valid role at the time the subscription is made.

- Since multiple subscribers may have different roles, and therefore different column access, the `realtime.apply_rls()` function now returns `setof realtime.wal_rls` instead of `realtime.wal_rls`

- The `Unauthorized` exception has been expanded from being raised when a role can not access any columns on a table to being raised if any of the primary key columns can not be selected by the role

- A `Bad Request` error was added and is thrown when the WAL record's table has no primary key (replaces #28)

- When errors are returned, the `subscription_ids` are return so users can be notified as needed

w3b6x9 commented on README.md:
````diff
@@ -159,15 +159,28 @@ Important Notes:
+```sql
+(
+    null,                            -- wal
+    null,                            -- is_rls_enabled
````
@olirice i think we should update the payload so that `is_rls_enabled` is always a bool instead of null. Realtime server strips this from broadcasted message but it's important when determining whether message should be broadcast to all or `subscriber_ids` array.

so it's never seen by any clients but `is_rls_enabled` is important to determine if all clients should receive this or just authorized users.

olirice replied:
sounds good. resolved in 177748c0745a6af476c96ce9118986976fe40230

w3b6x9 replied:
@olirice don't forget to update the README here so that `is_rls_enabled` is either true or false.

w3b6x9 commented on README.md:
````diff
@@ -159,15 +159,28 @@ Important Notes:
 ```sql
 (
     null,                            -- wal
     null,                            -- is_rls_enabled
````
@olirice see https://github.com/supabase/walrus/pull/30/files#r765372641

w3b6x9 replied:
@olirice don't forget to update the README here so that is_rls_enabled is either true or false.

w3b6x9 commented on README.md:
```diff
@@ -159,15 +159,28 @@ Important Notes:
+    null,                            -- wal
+    null,                            -- is_rls_enabled
+    [...],                              -- users,
+    array['Error 400: Bad Request, no primary key'] -- errors
```
@olirice just something to think about but what if a table has replica identity set to full? doesn't that technically mean the entire row becomes the key?

olirice replied:
kinda, but there's no enforcement of uniqueness (afaik). If there is no primary key we also lose the guarantee that the RLS checking prepared statement will hit an index.

Since disabling a primary key could be an optimization to improve insert throughput (high volume/large tables) it would be very risky to allow

w3b6x9 commented on sql/walrus--0.1.sql:
```diff
@@ -290,182 +292,218 @@ declare
-        visible_to_user_ids,
-        errors
-    )::realtime.wal_rls;
+    return next unnest(result_arr);
```
@oli this works in pg13 but breaks in pg14. Haven't tested pg12 yet.

w3b6x9 commented on sql/walrus--0.1.sql:
```diff
@@ -37,18 +36,22 @@ create type realtime.user_defined_filter as (
-
-    constraint pk_subscription primary key (id),
-    unique (entity, user_id, filters)
+    claims jsonb not null,
```
@olirice what about we default `claims` to `'{}'`?

olirice replied:
the insert trigger enforces that the claims contents contains a `role` key since its required, so the default would fail to insert

w3b6x9 replied:
@olirice

> anonymous users don't have a jwt

Realtime server receives JWT with the following claims for anonymous user subscribing to a topic:

```
{
  "role": "anon",
  "iat": 1625372973,
  "exp": 1940948973
}
```

So we should be fine adding in a `not null` constraint for `claims_role`.

> what do you think about adding a required parameter like `realtime.apply_rls(..., default_role regrole)`

I like this but the only problem is there's no way right now for devs to pass it in if they're using hosted Supabase and use a different anonymous role (not `anon`). We can of course add an option to the dashboard in the future but for now we can just tell devs to include `role` in JWT.

olirice replied:
ok awesome, thats a much easier change. The `claims_role` is now not null

w3b6x9 commented on sql/walrus--0.1.sql:
```diff
@@ -37,18 +36,22 @@ create type realtime.user_defined_filter as (
-    created_at timestamp not null default timezone('utc', now()),
-
-    constraint pk_subscription primary key (id),
-    unique (entity, user_id, filters)
```
@olirice

> -- Apply consistent order to filters so the unique constraint on
> -- (subscription_id, entity, filters) can't be tricked by a different filter order

should we add back a unique index here for (id, entity, filters)?

olirice replied:
The `id` alone is unique so it wouldn't help with duplicate subscriptions by a single user anymore.

We were previously able to do this because we had a reference to the `user_id`, that would be constant for multiple subscriptions

removed filter reordering here 71567088253a2c853a5adb3df4a1e774f28914fb

w3b6x9 replied:
@olirice ah, you're absolutely right!

w3b6x9 commented:
@olirice Wanted to point out these so far:

1. Making a change to a table that's under the publication but is not currently being subscribed to in `realtime.subscription` returns `null` values across the board when using the publication polling query. Not really an issue but in this scenario we should omit so it's one fewer row that needs to be processed by Realtime server.

2. I seem to be getting the following error when I've inserted two subscribers into `realtime.subscription` where one subscriber is anon role and the other is authenticated role and I try to get changes after inserting into a `public` schema table:

> ERROR:  query returned more than one row
> CONTEXT:  query: unnest(result_arr)
> PL/pgSQL function realtime.apply_rls(jsonb,integer) line 232 at RETURN NEXT

olirice commented:
> Making a change to a table that's under the publication but is not currently being subscribed to in realtime.subscription returns null values across the board when using the publication polling query. Not really an issue but in this scenario we should omit so it's one fewer row that needs to be processed by Realtime server.

good call. Updated in 5373c07 w/ tests

> I seem to be getting the following error when I've inserted two subscribers into realtime.subscription where one subscriber is anon role and the other is authenticated role and I try to get changes after inserting into a public schema table:

>ERROR: query returned more than one row
CONTEXT: query: unnest(result_arr)
PL/pgSQL function realtime.apply_rls(jsonb,integer) line 232 at RETURN NEXT

yeeep, I botched that pretty hard. Added a test and resolved in 8a79ee7

w3b6x9 commented on README.md:
```diff
@@ -159,15 +159,28 @@ Important Notes:
+(
+    null,                            -- wal
+    null,                            -- is_rls_enabled
+    [...],                              -- users,
```
@olirice replace `users` with `subscription_ids`. There are a few other places in README where `users` is still referenced.

olirice replied:
thanks, resolved

w3b6x9 commented on README.md:
```diff
@@ -159,15 +159,28 @@ Important Notes:
+
 ### Error 401: Unauthorized
-If a WAL record is passed through `realtime.apply_rls` and the `authenticated` role does not have permission to `select` any of the columns in that table, an `Unauthorized` error is returned with no WAL data.
+If a WAL record is passed through `realtime.apply_rls` and the `authenticated` role does not have permission to `select` the primary key columns in that table, an `Unauthorized` error is returned with no WAL data.
```
@olirice we should remove `authenticated` here b/c it can be any role now.

olirice replied:
fixed in be3133a

w3b6x9 commented on README.md:
```diff
@@ -285,16 +300,17 @@ from
     ) xyz
 where
     coalesce(pub.w2j_add_tables, '') <> ''
+    and array_length(xyz.subscription_ids, 1) > 0
```
@olirice what's the performance impact of `array_length` as the size of `subscription_ids` increases? Since we're checking to see if there's at least one `subscription_id` can't we just access `subscription_ids` array at the first index and check if it's not null?

olirice replied:
no harm in swapping it out. replaced in 364d1a9

w3b6x9 commented on sql/walrus--0.1.sql:
```diff
@@ -5,7 +5,6 @@
 
 create schema realtime;
 grant usage on schema realtime to postgres;
```
@olirice we can take this out too since `realtime` schema is created by `postgres`

olirice replied:
fixed in 96bb612

w3b6x9 commented on sql/walrus--0.1.sql:
```diff
@@ -113,7 +109,6 @@ create trigger tr_check_filters
 
 
 grant all on realtime.subscription to postgres;
```
@olirice we can take this out since `postgres` was responsible for create the schema and table.

olirice replied:
fixed in 96bb612

w3b6x9 commented on README.md:
```diff
@@ -24,12 +24,12 @@ User subscriptions are managed through a table
-    filters realtime.user_defined_filter[],
-    created_at timestamp not null default timezone('utc', now()),
-    constraint pk_subscription primary key (id)
+    filters realtime.user_defined_filter[] not null default '{}',
```
@olirice perhaps it's time to rename the type from `realtime.user_defined_filter` to something like `realtime.subscription_defined_filter`

olirice replied:
`user_defiend_filter` still seems right to me or `subscription_filter` could also work. If you have a preference lmk and i'll update

w3b6x9 commented on sql/walrus--0.1.sql:
```diff
@@ -290,182 +285,214 @@ declare
+                array['Error 400: Bad Request, no primary key']
+            )::realtime.wal_rls;
+
+        -- The claims role does not have SELECT permission to the primary key of entity
```
@olirice i think you're only checking select permission for primary key b/c it's more performant and check for the actual column is done in `realtime.subscription_check_filters`? 

Realtime server still sprays every change to every filter so we won't have great column security even after these changes. After this PR is in we should talk about better filter security.

olirice replied:
> why are you only checking select permission for primary key?

we only check the primary key here because the primary key cols are all that is necessary to check if the RLS passes.

- When a subscription is created we check to make sure that the subscriptions `claim_role` has access to the columns referenced by filters https://github.com/supabase/walrus/blob/generic_claims/sql/walrus--0.1.sql#L82
- When we're processing the WAL, we strip the columns the role shouldn't see from `record` and `old_record` keys https://github.com/supabase/walrus/blob/generic_claims/sql/walrus--0.1.sql#L424

w3b6x9 commented:
@olirice I'm having an issue where I have three subscribers in the `realtime.subscription` table and each has a different role (`anon`, `authenticated`, `service_role`) and when I listen to a table in the `public` schema  and the table has RLS disabled then only the `authenticated` user gets changes while the other two get back error `Error 401: Unauthorized`.

Here's the key parts of my setup:

```sql
create role anon                nologin noinherit;
create role authenticated       nologin noinherit;
create role service_role        nologin noinherit bypassrls;

grant usage                     on schema public to anon, authenticated, service_role;
alter default privileges in schema public grant all on tables to anon, authenticated, service_role;
alter default privileges in schema public grant all on functions to anon, authenticated, service_role;
alter default privileges in schema public grant all on sequences to anon, authenticated, service_role;

create publication supabase_realtime for all tables;
```

Create a table in `public` schema.

Add the three different roles to `realtime.subscription`.

Insert a change to the `public` schema table.

Get back three records after publication polling query but only the `authenticated` subscriber gets the changes while the other two get back errors.

olirice commented:
I was not able to reproduce that problem. Here's the complete setup:

```sql
create role anon                nologin noinherit;
create role authenticated       nologin noinherit;
create role service_role        nologin noinherit bypassrls;

grant usage                     on schema public to anon, authenticated, service_role;
alter default privileges in schema public grant all on tables to anon, authenticated, service_role;
alter default privileges in schema public grant all on functions to anon, authenticated, service_role;
alter default privileges in schema public grant all on sequences to anon, authenticated, service_role;

create publication supabase_realtime for all tables;

create table public.note(id int primary key);


insert into realtime.subscription(id, entity, claims)
values
	(extensions.uuid_generate_v4(), 'public.note', jsonb_build_object('role', 'authenticated')),
	(extensions.uuid_generate_v4(), 'public.note', jsonb_build_object('role', 'anon')),
	(extensions.uuid_generate_v4(), 'public.note', jsonb_build_object('role', 'service_role'));

select * from pg_create_logical_replication_slot('realtime', 'wal2json', false);

insert into public.note(id) values (1);


set search_path = '';

with pub as (
    select
        concat_ws(
            ',',
            case when bool_or(pubinsert) then 'insert' else null end,
            case when bool_or(pubupdate) then 'update' else null end,
            case when bool_or(pubdelete) then 'delete' else null end
        ) as w2j_actions,
        string_agg(realtime.quote_wal2json(format('%I.%I', schemaname, tablename)::regclass), ',') w2j_add_tables
    from
        pg_publication pp
        join pg_publication_tables ppt
            on pp.pubname = ppt.pubname
    where
        pp.pubname = 'supabase_realtime'
    group by
        pp.pubname
    limit 1
)
select
    xyz.wal,
    xyz.is_rls_enabled,
    xyz.subscription_ids,
    xyz.errors
from
    pub,
    lateral (
          select
            *
          from
             pg_logical_slot_get_changes(
                'realtime', null, null,
                'include-pk', '1',
                'include-transaction', 'false',
                'include-timestamp', 'true',
                'write-in-chunks', 'true',
                'format-version', '2',
                'actions', coalesce(pub.w2j_actions, ''),
                'add-tables', pub.w2j_add_tables
            )
    ) w2j,
    lateral (
        select
            x.wal,
            x.is_rls_enabled,
            x.subscription_ids,
            x.errors
        from
            realtime.apply_rls(
                wal := w2j.data::jsonb,
                max_record_bytes := 1048576
            ) x(wal, is_rls_enabled, subscription_ids, errors)
    ) xyz
where
    coalesce(pub.w2j_add_tables, '') <> ''
    and xyz.subscription_ids[1] is not null
```
<img width="779" alt="Screen Shot 2021-12-09 at 7 58 47 PM" src="https://user-images.githubusercontent.com/12958657/145504487-90a14fab-481c-4aa3-a6a4-ad5748293c30.png">


could you please post a complete min repro case?

w3b6x9 commented:
@olirice it works as expected after following your setup.

I got errors for `anon` and `service_role` b/c I created the `public` schema table first and then ran:

```sql
alter default privileges in schema public grant all on tables to anon, authenticated, service_role;
alter default privileges in schema public grant all on functions to anon, authenticated, service_role;
alter default privileges in schema public grant all on sequences to anon, authenticated, service_role;
```

After running the following I got it working:

```sql
GRANT ALL ON ALL TABLES IN SCHEMA public TO anon, authenticated, service_role;
GRANT ALL ON ALL SEQUENCES IN SCHEMA public TO anon, authenticated, service_role;
GRANT ALL ON ALL ROUTINES IN SCHEMA public TO anon, authenticated, service_role;
```

w3b6x9 reviewed (approved):
awesome work! lgtm!

TomasHubelbauer commented:
Hello, relating to supabase/supabase#4342 and my own discussion post here supabase/supabase#4742, is this PR going to enable listening to Realtime changes using the anon API key? And if so, how can one track the rollout of this feature now that it has been merged? Thank you

olirice commented:
> is this PR going to enable listening to Realtime changes using the anon API key?

yes it is

> how can one track the rollout of this feature

It merged into supabase/realtime here https://github.com/supabase/realtime/pull/209
its currently in QA and is estimated to roll out to GA in 1-2 weeks.

The most likely place to see activity between now and when it releases is https://github.com/supabase/supabase/issues/4842
